### PR TITLE
MAINT: Cleanup scipy.optimize.leastsq

### PR DESCRIPTION
**Pull request by twmr**
* Merge two 'if' blocks  in leastsq into one.
* Remove temporary variable 'mesg'.
* Introduce global LEASTSQ_SUCCESS list

### CODE REVIEW

**Comment by thisch** [comment]
The CI failure for [python 3.5](https://travis-ci.org/scipy/scipy/jobs/366517507) is most probably not related to this PR.

File: scipy/optimize/minpack.py [hunk 1]
```diff
@@ -452,9 +445,16 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
...
+        return (retval[0], cov_x) + retval[1:-1] + (errors[info][0], info)
     else:
-        return (retval[0], info)
+        if info not in LEASTSQ_SUCCESS:
+            if info > 0:
+                warnings.warn(errors[info][0], RuntimeWarning)
+            elif info == 0:
+                raise errors[info][1](errors[info][0])
+            else:
+                raise errors['unknown'][1](errors['unknown'][0])
```
Comment (thisch) [inline scipy/optimize/minpack.py]: If I understand the code in __minpack.h correctly `info` in this module/function can never be negative (see https://github.com/scipy/scipy/blob/11509c4a98edded6c59423ac44ca1b7f28fba1fd/scipy/optimize/__minpack.h#L597 and https://github.com/scipy/scipy/blob/11509c4a98edded6c59423ac44ca1b7f28fba1fd/scipy/optimize/__minpack.h#L499). Is this correct?

**Comment by thisch** [comment]
@pv, @WarrenWeckesser when you have time, can you please review this leastsq cleanup?

File: scipy/optimize/minpack.py [hunk 2]
```diff
@@ -452,9 +445,16 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
                 cov_x = inv(dot(transpose(R), R))
             except (LinAlgError, ValueError):
                 pass
-        return (retval[0], cov_x) + retval[1:-1] + (mesg, info)
+        return (retval[0], cov_x) + retval[1:-1] + (errors[info][0], info)
     else:
-        return (retval[0], info)
+        if info not in LEASTSQ_SUCCESS:
+            if info > 0:
```
Comment (rgommers) [inline scipy/optimize/minpack.py]: This should be `if info > [5, 6, 7, 8]`, as it was previously. Then the removal of the `try-except` is correct; here there's no guarantee that `info` will not be >8 it seems.

Comment (rgommers) [reply scipy/optimize/minpack.py]: You could use `LEASTSQ_FAILURE` for `[5, 6, 7, 8]`

**Comment by rgommers** [comment]
Thanks @thisch. This looks good to me, modulo my one minor comment.

File: scipy/optimize/minpack.py (hunk 2 above)
Comment (thisch) [reply scipy/optimize/minpack.py]: Thank you very much for taking the time to review my PR!

According to the fortran code `info` can't be >8. Therefore, I think that `info>0, info==0, info<0` checks in this PR makes the code a bit easier to understand, but I don't have a strong opinion on this. I'll introduce a new global variable `LEASTSQ_FAILURE` as you`ve suggested.

**Comment by thisch** [comment]
Thx for merging it!

**Pull request merged**